### DISCUSSION TITLE
Packaging Custom Models with Custom Functions Throws Error when Importing

### DISCUSSION THREAD

**Discussion by JRE-joe** [explosion/spaCy#7454] [Help: Coding & Implementations] [answered]
As an Example i'll be using the project rel_component from [https://github.com/explosion/projects/tree/v3/tutorials/rel_component](url) with the the original assets.

The first thing i have to change is in the ./scripts/custom_functions.py file so the custom scripts can be imported, otherwise there is an error when i debug my config, because the custom_functions can't be found. To solve this issue i edited:
```
import sys
sys.path.append('./scripts/')

# make the factory work
from rel_pipe import make_relation_extractor

# make the config work
from rel_model import create_relation_model, create_classification_layer, create_instances, create_tensors
```
Now a model can be trained. The goal is to get  installable python package of my trained model:
```
"python -m spacy package training/model-last packages --name ${vars.name} --version ${vars.version} --code ./scripts/custom_functions.py --force"
```
The built package can be installed in my virtual environment.
As soon as i import the model  ```import en_rel_sm``` there occurs the error: 
```
Exception has occurred: ModuleNotFoundError
No module named 'rel_pipe'
```
Looking in the package file, there is only the custom_functions.py file and the __init__.py file containing just the import for the custom_functions.py file. 
So at first i thought, due to the fact that it is a custom component with custom functions i have to include the rel_model.py and rel_pipe.py but this leads to the following error when packaging: ```"python -m spacy package training/model-last packages --name ${vars.name} --version ${vars.version} --code ./scripts/custom_functions.py,./scripts/rel_pipe.py,./scripts/rel_model.py --force" ```

Error:
```
================================== package ==================================
Running command: /Users/joe/Desktop/nlp/.env/bin/python3 -m spacy package training/model-last packages --name rel_sm --version 0.0.1 --code ./scripts/custom_functions.py,./scripts/rel_pipe.py,./scripts/rel_model.py --force
ℹ Building package artifacts: sdist
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/spacy/__main__.py", line 4, in <module>
    setup_cli()
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/spacy/cli/_util.py", line 69, in setup_cli
    command(prog_name=COMMAND)
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/typer/main.py", line 497, in wrapper
    return callback(**use_params)  # type: ignore
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/spacy/cli/package.py", line 56, in package_cli
    silent=False,
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/spacy/cli/package.py", line 92, in package
    util.import_file(code_path.stem, code_path)
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/spacy/util.py", line 1187, in import_file
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "scripts/rel_pipe.py", line 32, in <module>
    nlp: Language, name: str, model: Model, *, threshold: float
  File "/Users/joe/Desktop/nlp/.env/lib/python3.7/site-packages/spacy/language.py", line 470, in add_factory
    raise ValueError(err)
ValueError: [E004] Can't set up pipeline component: a factory for 'relation_extractor' already exists. Existing factory: <function make_relation_extractor at 0x7ff46e60c7a0>. New factory: <function make_relation_extractor at 0x7ff46e616440>
```

Is there any solution, packaging a trained model with the other python files (rel_pipe.py & rel_model.py) from this project so i can transfer the conclusions to my own project ?

_________________________________
System Setup:
Operating System: macOS Big Sur Version 11.2.2.
SpaCy: 3.0.5
python: 3.7.4
[...] other libraries

**Comment by JRE-joe** [comment, marked as answer]
The quick and dirty solution to this:
Packaging all 3 files together as one: custom_functions.py, rel_model.py and rel_pipe.py solves all the issues when importing.

Even though i found a solution to this, what could be the reason in SpaCy for all this ? I'll be interested if anybody knows an answer for future projects.

**Reply by svlandeg** [reply]
What happens when you run `spacy package ... ./scripts/rel_pipe.py,./scripts/rel_model.py`?

**Reply by JRE-joe** [reply]
the packaging process is sucessfull when i run 
```
"python -m spacy package training/model-last packages --name ${vars.name} --version ${vars.version} --code ./scripts/rel_pipe.py,./scripts/rel_model.py --force"
```
importing it into a script throws the error:
```
[E004] Can't set up pipeline component: a factory for 'relation_extractor' already exists. Existing factory: <function make_relation_extractor at 0x7ff7cb5dea70>. New factory: <function make_relation_extractor at 0x7ff7cb5e7680>
```

**Reply by svlandeg** [reply]
Sorry to hear that this is so frustrating, let me try and get to the bottom of this with you.

So if I understand correctly, the packaging now works succesfully. Next, you should be able to do 
```
pip install packages/${vars.name}-${vars.version}/dist/${vars.name}-${vars.version}.tar.gz
```
(if I got your parameters right - otherwise adjust the file name accordingly)

And when that is installed, you should be able to load your trained model with
```
nlp = spacy.load(name)
```
and this _should_ work without importing anything else in the same script - the custom functions should now all be available from the installed package...

Does that work?

If not, can you share the full script and error trace?
